src/utils/encoding/mod.rs

```diff
@@ -79,15 +83,123 @@ pub fn prover_id_from_u64(id: u64) -> ProverId {
     prover_id
 }
 
+// Unsigned VarInt (Uvi) Bytes
+pub struct UviBytes {
+    // cache for varint frame size
+    len: Option<usize>,
+    // content size limit, defaults to 128MiB
+    limit: usize,
+}
+
+impl UviBytes {
+    pub fn new_with_limit(limit: usize) -> Self {
+        UviBytes { len: None, limit }
+    }
+}
+
+impl Default for UviBytes {
+    fn default() -> Self {
+        UviBytes::new_with_limit(128 * 1024 * 1024)
+    }
+}
+
+impl Decoder for UviBytes {
+    type Item = BytesMut;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if self.len.is_none() {
+            if let Some((n, size)) = usize::decode_var(src) {
+                src.advance(size);
+                self.len = Some(n);
+            } else if src.len() >= std::mem::size_of::<usize>() {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "invalid uvi frame",
+                ));
+            }
+        }
+        if let Some(n) = self.len.take() {
+            if n > self.limit {
+                return Err(io::Error::new(
+                    io::ErrorKind::PermissionDenied,
+                    format!("uvi frame size limit exceeded, decode {}", n),
+                ));
+            }
+            if n <= src.len() {
+                return Ok(Some(src.split_to(n)));
+            }
+            src.reserve(n - src.len());
+            self.len = Some(n);
+        }
+        Ok(None)
+    }
+}
+
+impl Encoder<Bytes> for UviBytes {
+    type Error = io::Error;
+
+    fn encode(&mut self, item: Bytes, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        if item.remaining() > self.limit {
+            return Err(io::Error::new(
+                io::ErrorKind::PermissionDenied,
+                format!("uvi frame size limit exceeded, encode {}", item.remaining()),
+            ));
+        }
+        dst.reserve(item.remaining() + item.remaining().required_space());
+        let mut buffer = [0; 16];
+        let len = item.remaining().encode_var(&mut buffer);
+        dst.extend_from_slice(&buffer[0..len]);
```

Comment:
Could we avoid this? That is to say, write to `dst` directly?

Reply:
Nope. :(

Reply:
:crab: